saleor/discount/utils.py

```diff
@@ -345,17 +405,77 @@ def create_or_update_discount_objects_from_promotion_for_checkout(
     lines_info: Iterable["CheckoutLineInfo"],
 ):
     create_discount_objects_for_catalogue_promotions(lines_info)
-    create_discount_objects_for_order_promotions(checkout_info, lines_info)
+    create_checkout_discount_objects_for_order_promotions(checkout_info, lines_info)
+
+
+def create_or_update_discount_objects_from_promotion_for_order(
+    order: "Order",
+    lines_info: Iterable["DraftOrderLineInfo"],
+):
+    create_discount_objects_for_catalogue_promotions(lines_info)
+    _update_order_line_prefetched_discounts(lines_info)
+    _update_order_line_base_unit_prices(lines_info)
+
+    create_order_discount_objects_for_order_promotions(order, lines_info)
+    # if order promotion is type of gift, discount is associated with line, not order
+    _update_order_line_prefetched_discounts(lines_info)
+    _copy_unit_discount_data_to_order_line(lines_info)
```

Comment:
Then in this method, we can update the `line.discounts` and clear cache (if needed).

Reply:
Also I will suggest adding more comments explaining why we are doing those operations, as I suppose it will be hard to recall after a month